tests/unit/s2n_locking_test.c

```diff
@@ -0,0 +1,114 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+#include <pthread.h>
+
+#include "crypto/s2n_locking.h"
+
+#define LOCK_N 1
+
+#if !(S2N_OPENSSL_VERSION_AT_LEAST(1, 1, 0))
+
+static void* s2n_test_thread(void *arg)
+{
+    bool *passed_lock = (bool*) arg;
+    CRYPTO_lock(CRYPTO_LOCK, LOCK_N, NULL, 0);
+    *passed_lock = true;
+    CRYPTO_lock(CRYPTO_UNLOCK, LOCK_N, NULL, 0);
+    return NULL;
+}
+
+static void s2n_test_locking_cb(int mode, int n, char *file, int line)
+{
+    return;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* Test: libcrypto locks should actually lock.
+     *
+     * If libcrypto locking is not configured properly, then
+     * libcrypto locks will be no-ops and will not block threads.
+     */
+    {
+        /* Lock one of the libcrypto locks */
+        CRYPTO_lock(CRYPTO_LOCK, LOCK_N, NULL, 0);
+
+        /* Create a new thread which will try to take the lock held by this thread
+         * in order to set a flag.
+         */
+        bool passed_lock = false;
+        pthread_t thread = { 0 };
+        EXPECT_EQUAL(pthread_create(&thread, NULL, s2n_test_thread, &passed_lock), 0);
+
+        /* Expect the flag NOT to be set because we still hold the lock */
+        sleep(1);
+        EXPECT_FALSE(passed_lock);
+
+        /* Release the libcrypto lock */
+        CRYPTO_lock(CRYPTO_UNLOCK, LOCK_N, NULL, 0);
+
+        /* Expect the flag to be set because the new thread took the lock */
+        sleep(1);
+        EXPECT_TRUE(passed_lock);
+    }
+
+    /* Test: basic lifecycle
+     */
+    {
+        EXPECT_OK(s2n_locking_cleanup());
+        EXPECT_OK(s2n_locking_init());
+    }
+
+    /* Test: s2n-tls should not override locking configured by the application */
+    {
+        /* Manually override the existing callback. */
+        CRYPTO_set_locking_callback((void (*)()) s2n_test_locking_cb);
+        EXPECT_EQUAL(CRYPTO_get_locking_callback(), (void (*)()) s2n_test_locking_cb);
+
+        /* Cleaning up does not affect the application-set callback */
+        EXPECT_OK(s2n_locking_cleanup());
+        EXPECT_EQUAL(CRYPTO_get_locking_callback(), (void (*)()) s2n_test_locking_cb);
+
+        /* Initializing again doesn't override the application-set callback */
+        EXPECT_OK(s2n_locking_init());
+        EXPECT_EQUAL(CRYPTO_get_locking_callback(), (void (*)()) s2n_test_locking_cb);
+        EXPECT_OK(s2n_locking_cleanup());
+        EXPECT_EQUAL(CRYPTO_get_locking_callback(), (void (*)()) s2n_test_locking_cb);
+
+        /* Reset the callback */
+        CRYPTO_set_locking_callback(NULL);
+        EXPECT_EQUAL(CRYPTO_get_locking_callback(), NULL);
+
+        /* Initializing now sets the missing callback */
+        EXPECT_OK(s2n_locking_init());
+        EXPECT_NOT_EQUAL(CRYPTO_get_locking_callback(), NULL);
+    }
```

Comment:
Test that an application can override the callback set by s2n-tls? But I guess we implicitly actually do that in line 81.

Reply:
I added comments to try to make it clearer what's happening on line 81, and a check to verify that's actually what's  happening.